### PR TITLE
Fix bugs in explorer plugin

### LINKED ISSUE #1796: [BUG] Variable %LocalAppData% in path cause exception
I have backup folder contains **%LocalAppData%** in path. That variable is not used literally as %LocalAppData% but as absolute path.

Example
This is wrong

> E:\Backup\CloneApp\Backup\Windows Taskbar\C:\Users\account\AppData\Roaming\Microsoft\Internet Explorer\Quick Launch\User Pinned\TaskBar\TaskBar\Windows Sandbox.lnk

This is correct

> E:\Backup\CloneApp\Backup\Windows Taskbar\%AppData%\Microsoft\Internet Explorer\Quick Launch\User Pinned\TaskBar\TaskBar\Windows Sandbox.lnk

That cause exception and I can't find these files.

### LINKED ISSUE #1802: [BUG] - Explorer Plugin crash when first input is a colon ":"
Explorer Plugin immediately crash when this first input is a colon, this happen only when Explorer plugin is ON.
Same result with FR, US and ZH layout.

Not really a major issue, just wanted to let you know.

### Log file :
[2023-01-15.txt](https://github.com/Flow-Launcher/Flow.Launcher/files/10420173/2023-01-15.txt)

### Exception message :
Flow Launcher version: 1.11.0
OS Version: Microsoft Windows NT 10.0.19044.0
IntPtr Length: 8
x64: True

Python Path: C:\Python310\pythonw.exe
Node Path: 
Date: 01/15/2023 13:26:02
Exception:
Explorer Exception: 
Websites: https://github.com/Flow-Launcher/Flow.Launcher
Author: Jeremy Wu
Version: 2.1.0
Flow.Launcher.Core.ExternalPlugins.FlowPluginException: Object reference not set to an instance of an object.
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Flow.Launcher.Plugin.Explorer.Search.SearchManager.<>c__DisplayClass4_0.<SearchAsync>b__1(AccessLink excludedPath) in C:\projects\flow-launcher\Plugins\Flow.Launcher.Plugin.Explorer\Search\SearchManager.cs:line 120
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at System.Collections.Generic.HashSet`1.RemoveWhere(Predicate`1 match)
   at Flow.Launcher.Plugin.Explorer.Search.SearchManager.SearchAsync(Query query, CancellationToken token) in C:\projects\flow-launcher\Plugins\Flow.Launcher.Plugin.Explorer\Search\SearchManager.cs:line 119
   at Flow.Launcher.Plugin.Explorer.Main.QueryAsync(Query query, CancellationToken token) in C:\projects\flow-launcher\Plugins\Flow.Launcher.Plugin.Explorer\Main.cs:line 88
   at Flow.Launcher.Core.Plugin.PluginManager.<>c__DisplayClass22_0.<<QueryForPluginAsync>b__0>d.MoveNext() in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 191
--- End of stack trace from previous location ---
   at Flow.Launcher.Infrastructure.Stopwatch.DebugAsync(String message, Func`1 action) in C:\projects\flow-launcher\Flow.Launcher.Infrastructure\Stopwatch.cs:line 40
   at Flow.Launcher.Core.Plugin.PluginManager.QueryForPluginAsync(PluginPair pair, Query query, CancellationToken token) in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 213
   --- End of inner exception stack trace ---
   at Flow.Launcher.Core.Plugin.PluginManager.QueryForPluginAsync(PluginPair pair, Query query, CancellationToken token) in C:\projects\flow-launcher\Flow.Launcher.Core\Plugin\PluginManager.cs:line 213
   at Flow.Launcher.ViewModel.MainViewModel.<>c__DisplayClass135_0.<<QueryResults>g__QueryTask|2>d.MoveNext() in C:\projects\flow-launcher\Flow.Launcher\ViewModel\MainViewModel.cs:line 821
--- End of stack trace from previous location ---
   at Flow.Launcher.ViewModel.MainViewModel.QueryResults() in C:\projects\flow-launcher\Flow.Launcher\ViewModel\MainViewModel.cs:line 822
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)

### PR DESCRIPTION
**Pull request by VictoriousRaptor**
Changes:
- Fix #1802 (Fix null subtitle when creating disk result)
- Fix #1796 (Only expand environment var when path starts with %)
- Use case-insensitive comparator for path results
- Use api for subpath check

Tests:
- A single ":" when excluded path is not empty won't cause exception. (1802)
- Unit test for environment path expansion
- Unit test for subpath check
- Unit test for result comparator

### CODE REVIEW

**Comment by jjw24** [comment]
Please update description with what you have tested :)

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs [hunk 1]
```diff
@@ -44,40 +62,26 @@ internal static bool IsEnvironmentVariableSearch(string search)
...
-            if (splitSearch.Length == 2 && envStringPaths.ContainsKey(exactEnvStringPath))
-            {
-                var queryPartToReplace = $"%{exactEnvStringPath}%";
-                var expandedPath = envStringPaths[exactEnvStringPath];
-                // replace the %envstring% part of the query with its expanded equivalent
-                return environmentVariablePath.Replace(queryPartToReplace, expandedPath);
-            }
-
-            return environmentVariablePath;
+            return Environment.ExpandEnvironmentVariables(environmentVariablePath);
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs:65]: maybe just update TranslateEnvironmentVariablePath call with Environment.ExpandEnvironmentVariables? Doesnt look like it needs to be wrapped inside a method.

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs [hunk 2]
```diff
@@ -105,19 +107,21 @@ when ActionKeywordMatch(query, Settings.ActionKeyword.IndexSearchActionKeyword)
...
-                if (e is EngineNotAvailableException)
-                    throw;
-
                 throw new SearchException(engineName, e.Message, e);
             }
-            
+
             results.RemoveWhere(r => Settings.IndexSearchExcludedSubdirectoryPaths.Any(
-                excludedPath => r.SubTitle.StartsWith(excludedPath.Path, StringComparison.OrdinalIgnoreCase)));
+                excludedPath => IsSubPathOf(r.SubTitle, excludedPath.Path)));
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:124]: what's the issue with this check?

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs [hunk 3]
```diff
@@ -8,14 +8,15 @@
...
+using System.IO;
 
 namespace Flow.Launcher.Plugin.Explorer.Search
 {
     public class SearchManager
     {
-        internal static PluginInitContext Context;
+        internal PluginInitContext Context;
 
-        internal static Settings Settings;
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:18]: Why are we changing this from static?

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs [hunk 4]
```diff
@@ -30,12 +31,13 @@ private class PathEqualityComparator : IEqualityComparer<Result>
...
             {
-                return x.Title == y.Title && x.SubTitle == y.SubTitle;
+                return x.Title.Equals(y.Title, StringComparison.OrdinalIgnoreCase) 
+                    && string.Equals(x.SubTitle, y.SubTitle, StringComparison.OrdinalIgnoreCase);
             }
 
             public int GetHashCode(Result obj)
             {
-                return HashCode.Combine(obj.Title.GetHashCode(), obj.SubTitle?.GetHashCode() ?? 0);
+                return HashCode.Combine(obj.Title.ToLowerInvariant(), obj.SubTitle?.ToLowerInvariant() ?? "");
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:40]: Is the GetHashCode not required? 

Why is the return type from the null check changed?

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs (hunk 1 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs:65]: Sure.

**Comment by jjw24** [comment]
@z1nc0r3 just a FYI the issue with ':' is addressed here already.

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs (hunk 2 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:124]: a path string starts with another doesn't necessarily mean it's contains the other one (unless ending with '\' is guaranteed). like `c:\program files` and `c:\program`

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs (hunk 3 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:18]: search manager is Singleton but with these static members initialized in constructor. imo it's confusing

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs [hunk 5]
```diff
@@ -252,5 +256,16 @@ internal static bool IsEnvironmentVariableSearch(string search)
                    && search != "%%"
                    && !search.Contains('\\');
         }
+
+        // https://stackoverflow.com/a/66877016
+        internal static bool IsSubPathOf(string subPath, string basePath)
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:261]: Please move this method to FilesFolders class, which can then be shared by other plugins and flow.

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs (hunk 4 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:40]: it's not required. using hashcode.combine is to generating hash code based on multiple values, not to hash their hash code again.

**Comment by VictoriousRaptor** [comment]
> @z1nc0r3 just a FYI the issue with ':' is addressed here already.

that's a bit different. that pr was to ignore queries with ':'. this one is to fix the crash issue.

**Comment by jjw24** [comment]
The only other thing I picked up is the use of %USERNAME% in the form of `c:\users\%USERNAME%\downloads`, I dont know how many other variables that used during mid-path.

File: Flow.Launcher.Test/FilesFoldersTest.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace Flow.Launcher.Test
+{
+    [TestFixture]
+    
+    public class FilesFoldersTest
+    {
+        // Testcases from https://stackoverflow.com/a/31941905/20703207
+        // Disk
+        [TestCase(@"c:", @"c:\foo", true)]
```
Comment (jjw24) [inline Flow.Launcher.Test/FilesFoldersTest.cs:12]: This failed the test. Using `c:` works in Win Explorer but not supported in Explorer plugin, is this test still applicable?

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs [hunk 7]
```diff
@@ -8,12 +8,30 @@ namespace Flow.Launcher.Plugin.Explorer.Search
...
+        private static Dictionary<string, string> _envStringPaths = null;
+        private static Dictionary<string, string> EnvStringPaths
+        {
+            get
+            {
+                if (_envStringPaths == null)
+                {
+                    _envStringPaths = LoadEnvironmentStringPaths();
+                }
+                return _envStringPaths;
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs:20]:
```suggestion
                return _envStringPaths ?? LoadEnvironmentStringPaths();
```

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs [hunk 8]
```diff
@@ -13,29 +13,33 @@ namespace Flow.Launcher.Plugin.Explorer.Search
...
 
         public SearchManager(Settings settings, PluginInitContext context)
         {
             Context = context;
             Settings = settings;
         }
 
-        private class PathEqualityComparator : IEqualityComparer<Result>
+        /// <summary>
+        /// Note: Assuming all diretories end with "\".
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:27]:
```suggestion
        /// Note: A path that ends with "\" and one that doesn't will not be regarded as equal.
```

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs [hunk 9]
```diff
@@ -105,19 +109,21 @@ when ActionKeywordMatch(query, Settings.ActionKeyword.IndexSearchActionKeyword)
...
+                throw;
+            }
             catch (Exception e)
             {
-                if (e is OperationCanceledException)
-                    return results.ToList();
-
-                if (e is EngineNotAvailableException)
-                    throw;
-
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:115]: Why this change?

Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:115]: Just for clarity.

**Comment by VictoriousRaptor** [comment]
> The only other thing I picked up is the use of %USERNAME% in the form of `c:\users\%USERNAME%\downloads`. https://www.elevenforum.com/t/complete-list-of-environment-variables-in-windows-11.11212/ I suppose though you can use %USERPROFILE% to achieve the same as using %USERNAME%
> 
> Perhaps add an exclusion for %USERNAME%?

Just found something interesting. If you create a folder called `%appdata%` in some folder like `D:\foo`, then try to cd `D:\foo\%appdata%` in cmd or type it in the location bar of windows explorer, it doesn't work. But you can still enter it by double clicking. Guess we need to reconsider how to deal withi it.

File: Flow.Launcher.Test/FilesFoldersTest.cs (hunk 6 above)
Comment (VictoriousRaptor) [reply Flow.Launcher.Test/FilesFoldersTest.cs:12]: It passes my local test. Weird. Should I disable it for CI?

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs [hunk 10]
```diff
@@ -1,83 +1,88 @@
...
         }
 
-        internal static Dictionary<string, string> LoadEnvironmentStringPaths()
+        public static bool HasEnvironmentVar(string search)
         {
-            var envStringPaths = new Dictionary<string, string>(StringComparer.InvariantCultureIgnoreCase);
+            // "c:\foo %appdata%\" returns false
+            var splited = search.Split(Path.DirectorySeparatorChar);
+            return splited.Any(dir => dir.StartsWith('%') && 
+                                        dir.EndsWith('%') &&
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs:38]: I don't think we need endswith? Some may like to write `%appdata%\flowlauncher\` balblabla. No need to enforce expansion to go deeper

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs [hunk 11]
```diff
@@ -105,19 +109,21 @@ when ActionKeywordMatch(query, Settings.ActionKeyword.IndexSearchActionKeyword)
                 await foreach (var search in searchResults.WithCancellation(token).ConfigureAwait(false))
                     results.Add(ResultManager.CreateResult(query, search));
             }
+            catch (OperationCanceledException)
+            {
+                return results.ToList();
```
Comment (taooceros) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/SearchManager.cs:114]: Just return an empty list here. Don't bother the Tolist

**Comment by taooceros** [comment]
> > The only other thing I picked up is the use of %USERNAME% in the form of `c:\users\%USERNAME%\downloads`. https://www.elevenforum.com/t/complete-list-of-environment-variables-in-windows-11.11212/ I suppose though you can use %USERPROFILE% to achieve the same as using %USERNAME%
> > Perhaps add an exclusion for %USERNAME%?
> 
> Just found something interesting. If you create a folder called `%appdata%` in some folder like `D:\foo`, then try to cd `D:\foo\%appdata%` in cmd or type it in the location bar of windows explorer, it doesn't work. But you can still enter it by double clicking. Guess we need to reconsider how to deal with it.
> 
> Edit: Only expand when matching `...\%blabla%\...`.

What about check whether the path exists before expanding environmental variable?

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs (hunk 10 above)
Comment (VictoriousRaptor) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs:38]: It matches spilt path (%appdata% and flowlauncher), not the whole path.

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/EnvironmentVariables.cs:38]: oh sorry I read it wrongly thanks

**Comment by VictoriousRaptor** [comment]
> > > The only other thing I picked up is the use of %USERNAME% in the form of `c:\users\%USERNAME%\downloads`. https://www.elevenforum.com/t/complete-list-of-environment-variables-in-windows-11.11212/ I suppose though you can use %USERPROFILE% to achieve the same as using %USERNAME%
> > > Perhaps add an exclusion for %USERNAME%?
> > 
> > 
> > Just found something interesting. If you create a folder called `%appdata%` in some folder like `D:\foo`, then try to cd `D:\foo\%appdata%` in cmd or type it in the location bar of windows explorer, it doesn't work. But you can still enter it by double clicking. Guess we need to reconsider how to deal with it.
> > Edit: Only expand when matching `...\%blabla%\...`.
> 
> What about check whether the path exists before expanding environmental variable?

For relative path like %homepath% (relative to %homedrive%) we need to expand it before checking existence.

**Comment by VictoriousRaptor** [comment]
> What about check whether the path exists before expanding environmental variable?

Expanding is before checking the path is incomplete (user inputting and without trailing '\\'). For this bug this pr is enough imo. Even Windows Explorer can't deal with this situation perfectly.

**Comment by taooceros** [comment]
> Expanding is before checking the path is incomplete (user inputting and without trailing '\'). For this bug this pr is enough imo. Even Windows Explorer can't deal with this situation perfectly.

lol ok

**Comment by VictoriousRaptor** [comment]
> > Expanding is before checking the path is incomplete (user inputting and without trailing ''). For this bug this pr is enough imo. Even Windows Explorer can't deal with this situation perfectly.
> 
> lol ok

It can be done but I don't really want to do it. In this pr @kubalav 's path should be correctly parsed. but paths like d:\%appdata% still can't be handled correctly.

File: Flow.Launcher.Plugin/SharedCommands/FilesFolders.cs [hunk 12]
```diff
@@ -1,7 +1,9 @@
 ﻿using System;
 using System.Diagnostics;
 using System.IO;
+#if !DEBUG
```
Comment (taooceros) [inline Flow.Launcher.Plugin/SharedCommands/FilesFolders.cs:4]: why this?

Comment (VictoriousRaptor) [reply Flow.Launcher.Plugin/SharedCommands/FilesFolders.cs:4]: It was to neutralize warning. `System.Windows` is only used for release. Just realized I should neutralize the warning instead.

File: Flow.Launcher.Plugin/SharedCommands/FilesFolders.cs [hunk 13]
```diff
@@ -241,5 +237,33 @@ public static string ReturnPreviousDirectoryIfIncompleteString(string path)
...
+                   && !rel.StartsWith(@"..\")
+                   && !Path.IsPathRooted(rel);
+        }
+        
+        /// <summary>
+        /// Returns path ended with "\"
+        /// </summary>
+        /// <param name="path"></param>
+        /// <returns></returns>
+        public static string EnsureTrailingSlash(this string path)
```
Comment (jjw24) [inline Flow.Launcher.Plugin/SharedCommands/FilesFolders.cs:264]: nice, this will be helpful because through out explorer there are dup code to add trailing slash.

**Review by jjw24 — APPROVED** [review]
